metrics/sources/kubelet/summary.go

```diff
@@ -0,0 +1,342 @@
+
+	// Need to translate from nanocores to millicores.
+	if cpu.UsageNanoCores != nil {
+		const nanoPerMilli uint64 = 1e6
+		cores := *cpu.UsageNanoCores / nanoPerMilli
+		this.addIntMetric(metrics, &MetricCpuUsageRate, &cores, startTime)
+	}
+}
+
+func (this *summaryMetricsSource) decodeMemoryStats(metrics *MetricSet, memory *stats.MemoryStats, startTime time.Time) {
+	if memory == nil {
+		return
+	}
+
+	this.addIntMetric(metrics, &MetricMemoruUsage, memory.UsageBytes, startTime)
+	this.addIntMetric(metrics, &MetricMemoryWorkingSet, memory.WorkingSetBytes, startTime)
+	this.addIntMetric(metrics, &MetricMemoryPageFaults, memory.PageFaults, startTime)
+	this.addIntMetric(metrics, &MetricMemoryMajorPageFaults, memory.MajorPageFaults, startTime)
+}
+
+func (this *summaryMetricsSource) decodeNetworkStats(metrics *MetricSet, network *stats.NetworkStats, startTime time.Time) {
+	if network == nil {
+		return
+	}
+
+	this.addIntMetric(metrics, &MetricNetworkRx, network.RxBytes, startTime)
+	this.addIntMetric(metrics, &MetricNetworkRxErrors, network.RxErrors, startTime)
+	this.addIntMetric(metrics, &MetricNetworkTx, network.TxBytes, startTime)
+	this.addIntMetric(metrics, &MetricNetworkTxErrors, network.TxErrors, startTime)
+}
+
+func (this *summaryMetricsSource) decodeFsStats(metrics *MetricSet, fsKey string, fs *stats.FsStats) {
+	if fs == nil {
+		return
+	}
+
+	fsLabels := map[string]string{LabelResourceID.Key: fsKey}
+	if fs.UsedBytes != nil {
+		usage := LabeledMetric{
+			Name:       MetricFilesystemUsage.Name,
+			Labels:     fsLabels,
+			ValueType:  ValueInt64,
+			MetricType: MetricFilesystemUsage.Type,
+			IntValue:   int64(*fs.UsedBytes),
+		}
+		metrics.LabeledMetrics = append(metrics.LabeledMetrics, usage)
+	}
+
+	if fs.CapacityBytes != nil {
+		limit := LabeledMetric{
+			Name:       MetricFilesystemLimit.Name,
+			Labels:     fsLabels,
+			ValueType:  ValueInt64,
+			MetricType: MetricFilesystemLimit.Type,
+			IntValue:   int64(*fs.CapacityBytes),
+		}
+		metrics.LabeledMetrics = append(metrics.LabeledMetrics, limit)
+	}
+}
+
+func (this *summaryMetricsSource) decodeUserDefinedMetrics(metrics *MetricSet, udm []stats.UserDefinedMetric) {
+	for _, metric := range udm {
+		mv := MetricValue{}
+		switch metric.Type {
+		case stats.MetricGauge:
+			mv.MetricType = MetricGauge
+		case stats.MetricCumulative:
+			mv.MetricType = MetricCumulative
+		default:
+			glog.V(4).Infof("Skipping %s: unknown custom metric type: %v", metric.Name, metric.Type)
+			continue
+		}
+
+		const epsilon = 1e-9 // Margin of error
+		_, frac := math.Modf(math.Abs(metric.Value))
+		if metric.Value < math.MaxInt64 && (frac < epsilon || frac > 1.0-epsilon) {
+			mv.ValueType = ValueInt64
+			mv.IntValue = int64(metric.Value)
+			if frac > 1.0-epsilon {
+				mv.IntValue = mv.IntValue + 1 // Round up.
+			}
+		} else {
+			mv.ValueType = ValueFloat
+			mv.FloatValue = float32(metric.Value)
```

Comment:
Because only float32 values are supported, which only have 23 bits of precision. The integer conversion is an attempt to offer higher precision for integer values (such as counts). Do you think this is overcomplicating the conversion?